core/app/finders/spree/products/find.rb

```diff
@@ -324,6 +341,18 @@ def order_by_price(scope, sort_order)
       def order_by_best_selling(scope)
         scope.by_best_selling(:desc)
       end
+
+      def products_matching_all_taxonomies_ids(products_ids, taxon_groups)
+        classifications = Spree::Classification.
+                          where(product_id: products_ids, taxon_id: taxon_groups.flatten).
+                          pluck(:product_id, :taxon_id)
+
+        # Find products ids that match all taxonomies to tighten filter results
+        products_ids.select do |product_id|
+          product_taxon_ids = classifications.select { |product, _| product == product_id }.map { |_, taxon_id| taxon_id.to_s }
+          taxon_groups.all? { |group| (group & product_taxon_ids).any? }
+        end
```

Comment:
Could we cache product IDs with their associated taxon IDs?

```suggestion
        classifications = Spree::Classification.
          where(product_id: products_ids, taxon_id: taxon_groups.flatten).
          select(:product_id, :taxon_id).
          group_by(&:product_id).
          transform_values { |values| values.map(&:taxon_id).map(&:to_s) }

        # Find products ids that match all taxonomies to tighten filter results
        products_ids.select do |product_id|
          product_taxon_ids = classifications[product_id]
          taxon_groups.all? { |group| (group & product_taxon_ids).any? }
        end
```

Reply:
@caiodsc i did suggested changes with small change. STRING_AGG only wokrs with string types so i needed to add `Arel.sql("STRING_AGG(taxon_id::text, ',')"))`. Apart of that everything is working fine.

Reply:
@brozek95 if it makes sense for you:
```
# Replace:
product_id if taxon_groups.all? { |group| (group & product_taxon_ids).any? }
# With:
product_id if taxon_groups.all? { |group| group.intersect?(product_taxon_ids) }
```
`intersect?` will be a bit more efficient because it returns a boolean directly, whereas `&` creates a new array and then checks for `any?`, using more memory.

Doc: https://docs.ruby-lang.org/en/3.3/Array.html#method-i-intersect-3F.
If you're curious, you can also check the source code of the method to compare implementations.

This was my last suggestion... sorry if I'm being too nitpicky or blocking the PR too much 🙏

Reply:
Done, good idea, thanks.